### PR TITLE
Making SetVertex/SetEdge add elements in single transaction

### PR DESCRIPTION
**Pull request by kellrott**
KVGraph was adding vertices and edges one transaction at a time. This speeds things up.

### CODE REVIEW

File: kvgraph/kvgraph.go [hunk 1]
```diff
@@ -85,52 +85,58 @@ func (kgdb *KVInterfaceGDB) GetTimestamp() string {
...
 func (kgdb *KVInterfaceGDB) SetEdge(edgeArray []*aql.Edge) error {
-	for _, edge := range edgeArray {
-		if edge.Gid == "" {
-			eid := fmt.Sprintf("%d", rand.Uint64())
-			for ; kgdb.kv.HasKey(EdgeKeyPrefix(kgdb.graph, eid)); eid = fmt.Sprintf("%d", rand.Uint64()) {
+	kgdb.kv.Update(func(tx KVTransaction) error {
+		for _, edge := range edgeArray {
+			if edge.Gid == "" {
+				eid := fmt.Sprintf("%d", rand.Uint64())
+				for ; tx.HasKey(EdgeKeyPrefix(kgdb.graph, eid)); eid = fmt.Sprintf("%d", rand.Uint64()) {
```
Comment (buchanae) [inline kvgraph/kvgraph.go]: What's going on here? This is cryptic.

File: kvgraph/kvgraph.go [hunk 2]
```diff
@@ -85,52 +85,58 @@ func (kgdb *KVInterfaceGDB) GetTimestamp() string {
...
+		for _, vertex := range vertexArray {
+			d, _ := proto.Marshal(vertex)
+			k := VertexKey(kgdb.graph, vertex.Gid)
+			err := tx.Set(k, d)
+			if err != nil {
+				return err
+			}
+			kgdb.ts.Touch(kgdb.graph)
 		}
-		kgdb.ts.Touch(kgdb.graph)
```
Comment (buchanae) [inline kvgraph/kvgraph.go]: Why not one touch at the end?

File: kvgraph/interface.go [hunk 3]
```diff
@@ -34,9 +34,11 @@ type KVIterator interface {
 	Get(key []byte) ([]byte, error)
 }
 
-// KVTransaction is a genetic interface used by KVInterface.Update to allow the
```
Comment (buchanae) [inline kvgraph/interface.go]: haha

File: kvgraph/kvgraph.go (hunk 1 above)
Comment (kellrott) [reply kvgraph/kvgraph.go]: Random key assignment

File: kvgraph/kvgraph.go [hunk 4]
```diff
@@ -85,52 +85,58 @@ func (kgdb *KVInterfaceGDB) GetTimestamp() string {
...
-		}
-		err = kgdb.kv.Set(skey, []byte{})
-		if err != nil {
-			return err
-		}
-		err = kgdb.kv.Set(dkey, []byte{})
-		if err != nil {
-			return err
+			eid := edge.Gid
+			data, _ := proto.Marshal(edge)
```
Comment (buchanae) [inline kvgraph/kvgraph.go]: Save yourself the trouble and return all errors now.

File: kvgraph/kvgraph.go (hunk 1 above)
Comment (buchanae) [reply kvgraph/kvgraph.go]: A helper function called "randomKeyAssignment" would be easy.

**Comment by buchanae** [comment]
👍

**Pull request merged**